MF-795: Add swr to vitals widget

denniskigen opened the pull request:
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [x] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

Apologies for the huge diff.

This PR:

1. Adds SWR data fetching hooks to the Vitals widget.
2. Refactors the existing components to reuse these hooks.
3. Refactors existing tests to work within the SWR paradigm.
4. Moves `useVitalsConceptMetadata` (a hook for fetching vitals and biometrics concept units and metadata) to `esm-patient-common-lib` so it can be reused by both the vitals and biometrics widgets. Also moves the `withUnit` helper function which renders a label for a vital sign along with its associated unit.
5. Adds a mutate call in the VitalsandBiometricsForm so that the related widgets automatically revalidate when a new set of vitals or biometrics is successfully recorded via the form.

## Issue

https://issues.openmrs.org/browse/MF-795

brandones commented on packages/esm-patient-vitals-app/src/vitals/vitals-header/vitals-header-title.component.tsx:
```diff
@@ -1,24 +1,24 @@
 }
 
-const VitalsHeaderStateTitle: React.FC<VitalsHeaderStateTitleProps> = ({
+const VitalsHeaderTitle: React.FC<VitalsHeaderTitleProps> = ({
```
This is still a funny name. Is it a header, or a title? Or a header tile? Or the title of the header, because for some reason the body has its own title, which is different??

brandones commented on packages/esm-patient-vitals-app/src/vitals/vitals.resource.tsx:
```diff
@@ -1,81 +1,78 @@
+
+  const observations = data?.data?.total > 0 ? data.data?.entry?.map((entry) => entry.resource ?? []) : null;
+
+  const systolicBloodPressureData = observations?.filter((obs) =>
+    obs.code.coding.some((sys) => sys.code === config.concepts.systolicBloodPressureUuid),
+  );
+  const diastolicBloodPressureData = observations?.filter((obs) =>
+    obs.code.coding.some((sys) => sys.code === config.concepts.diastolicBloodPressureUuid),
+  );
+  const pulseData = observations?.filter((obs) =>
+    obs.code.coding.some((sys) => sys.code === config.concepts.pulseUuid),
+  );
+  const temperatureData = observations?.filter((obs) =>
+    obs.code.coding.some((sys) => sys.code === config.concepts.temperatureUuid),
   );
+  const oxygenSaturationData = observations?.filter((obs) =>
+    obs.code.coding.some((sys) => sys.code === config.concepts.oxygenSaturationUuid),
+  );
+  const heightData = observations?.filter((obs) =>
+    obs.code.coding.some((sys) => sys.code === config.concepts.heightUuid),
+  );
+  const weightData = observations?.filter((obs) =>
+    obs.code.coding.some((sys) => sys.code === config.concepts.weightUuid),
+  );
+  const respiratoryRateData = observations?.filter((obs) =>
+    obs.code.coding.some((sys) => sys.code === config.concepts.respiratoryRateUuid),
+  );
```
I think this was better as it was before. That `filterByConceptUuid` function could even be defined as an arrow function right on top of these lines, so it's all right there.

brandones commented on packages/esm-patient-vitals-app/src/vitals/vitals-overview.component.tsx:
```diff
@@ -26,84 +25,67 @@ const VitalsOverview: React.FC<VitalsOverviewProps> = ({ patientUuid, showAddVit
     {
       key: 'bloodPressure',
-      header: withUnit('BP', bloodPressureUnit),
+      header: withUnit('BP', conceptUnits ? conceptUnits[0] : ''),
```
```suggestion
      header: withUnit('BP', conceptUnits[0] ?? ''),
```

Right? Also I wonder if `withUnit` should just treat `undefined` like `''`.

brandones commented on packages/esm-patient-vitals-app/src/vitals/vitals-header/vitals-header.component.tsx:
```diff
@@ -0,0 +1,102 @@
+              <div className={styles.row}>
+                <VitalsHeaderItem
+                  unitName={t('temperatureAbbreviated', 'Temp')}
+                  unitSymbol={conceptUnits ? conceptUnits[2] : ''}
```
At some point (not in this PR) I think the vitals will need some refactoring to make it more robust, especially once we want to make it configurable. e.g. having these really arbitrary indexes tied to specific vitals is not awesome.

brandones reviewed (approved):
A few small things, nothing major. I think the major thing is that all this vitals stuff will need some refactoring in the near future. And the situation with the units is bad.

Nice work, generally

brandones commented:
Feel free to resolve conflicts, address feedback at your discretion, and merge.

denniskigen commented:
Thanks, @brandones, for your review and suggestions. Unfortunately, the scope of this PR ended up far exceeding its original intention. I agree about refactoring being required very shortly - especially around the units stuff. The latest commit addresses the issues around the use of `withUnit` and the `useVitals` hook. The remaining issues should be resolved outside of this PR, as you suggest. I'll proceed to merge this.
